test/sqllogictest/transform/predicate_reduction.slt

```diff
@@ -0,0 +1,96 @@
+# Copyright Materialize, Inc. All rights reserved.
+#
+# Use of this software is governed by the Business Source License
+# included in the LICENSE file at the root of this repository.
+#
+# As of the Change Date specified in that file, in accordance with
+# the Business Source License, use of this software will be governed
+# by the Apache License, Version 2.0.
+
+# This file contains tests for simplification of predicates.
+
+statement ok
+CREATE TABLE t1(f1 int, f2 int)
+
+statement ok
+INSERT INTO t1 VALUES (null, null), (0, null), (1, null), (1, 0), (null, 0)
+
+# Currently, we can simplify predicates with overlapping predicate
+# subexpressions only if the subexpression cannot be null.
+query II
+SELECT * FROM t1 WHERE f1 = 0 and (f1 = 0 or f1 = 1)
+----
+0
+NULL
+
+query T multiline
+EXPLAIN SELECT * FROM t1 WHERE f1 = 0 and (f1 = 0 or f1 = 1)
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter ((#0 = 0) || (#0 = 1)), (#0 = 0)
+
+EOF
+
+query T multiline
+EXPLAIN SELECT * FROM t1 WHERE f1 is null and (f1 is null or f1 = 1)
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter isnull(#0)
+
+EOF
+
+query II
+SELECT * FROM t1 WHERE f1 is null and (f1 is null or f1 = 1)
+----
+NULL
+NULL
+NULL
+0
+
+# Test that subexpression matching can detect a `!(predicate)` and then replace
+# other instances of `predicate` with `false`.
+
+query T multiline
+EXPLAIN SELECT * FROM t1 WHERE f1 is not null and (f1 is null or f1 = 1)
+----
+%0 =
+| Get materialize.public.t1 (u1)
+| Filter !(isnull(#0)), (#0 = 1)
+
+EOF
+
+query II
+SELECT * FROM t1 WHERE f1 is not null and (f1 is null or f1 = 1)
+----
+1
+NULL
+1
+0
+
+# A test that simplification works when overlapping subexpressions are nested.
```

Comment:
This query is correctly simplified:

```
CREATE TABLE t1 (col_null INTEGER, col_not_null INTEGER NOT NULL);

materialize=> EXPLAIN SELECT * FROM t1 WHERE (col_not_null + 2) = 1 AND (((col_not_null + 2) = 1) = true);
           Optimized Plan           
------------------------------------
 %0 =                              +
 | Get materialize.public.t1 (u233)+
 | Filter (1 = (#1 + 2))           +
```

But this one, which differs only in the arithmetic operator used, is not:

```
materialize=> EXPLAIN SELECT * FROM t1 WHERE (col_not_null % 2) = 1 AND (((col_not_null % 2) = 1) = true);
                  Optimized Plan                  
--------------------------------------------------
 %0 =                                            +
 | Get materialize.public.t1 (u233)              +
 | Filter (true = (1 = (#1 % 2))), (1 = (#1 % 2))+
```

Reply:
> But this one, which differs only in the arithmetic operator used, is not:
> 
> ```
> materialize=> EXPLAIN SELECT * FROM t1 WHERE (col_not_null % 2) = 1 AND (((col_not_null % 2) = 1) = true);
>                   Optimized Plan                  
> --------------------------------------------------
>  %0 =                                            +
>  | Get materialize.public.t1 (u233)              +
>  | Filter (true = (1 = (#1 % 2))), (1 = (#1 % 2))+
> ```

I looked at the code determining nullability, and it seems that the cause of the lack of simplification is as follows:
We used to return null when there is an evaluation error. So, `select * from t1.f1/t1.f2 > 0` would return null if any row had `t1.f2 = 0`. We have since changed the behavior so we return to an evaluation error instead, but we forgot to change the code determining whether the predicate can return null in light of this.
We were also very lazy in determining if a predicate can be null, so if the function is `%`, then we automatically return that the predicate can be null even if the second argument is a non-zero literal.